Automattic/wp-calypso
# Optimize loading of moment data

This pull request seeks to fix https://github.com/Automattic/wp-calypso/issues/12313 in order to reduce the size of our `vendor` bundle by trimming unnecessary`moment` data.

## Problem

We use the [`moment` module](https://www.npmjs.com/package/moment) in Calypso, either [directly](https://github.com/Automattic/wp-calypso/blob/9171888c9fe8890dac7d5bf81dc958d8db47017c/client/my-sites/upgrades/components/domain-warnings/index.jsx#L6) or [through](https://github.com/Automattic/wp-calypso/blob/9171888c9fe8890dac7d5bf81dc958d8db47017c/client/me/purchases/remove-purchase/index.jsx#L8) the [`i18n-calypso` module](https://www.npmjs.com/package/i18n-calypso). The latter actually [exposes an instance](https://github.com/Automattic/i18n-calypso/#moment-method) of the [`moment-timezone` module](https://www.npmjs.com/package/moment-timezone), which is just [an instance](https://github.com/moment/moment-timezone/blob/de5d1f8d17074af873366b2453b442552ddd35e4/moment-timezone.js#L18) of `moment` with additional [timezone features](http://momentjs.com/timezone/).

The problem is that both `moment` and `moment-timezone` comes with lots of data that either aren't really necessary or don't have to be loaded right away, but are included in the `vendor` chunk - because of how Webpack bundles modules. To be more specific, Webpack notices that `moment` (potentially) [depends on all of the locales](https://github.com/moment/moment/blob/b8a7fc310eb3625e83fc0c8f1ea2840fa83c7378/moment.js#L1830) - which represents [552KB of data](https://github.com/moment/moment/tree/b8a7fc310eb3625e83fc0c8f1ea2840fa83c7378/locale) - and bundles them. Same thing with `moment-timezone` that loads [176 KB of data](https://github.com/moment/moment-timezone/blob/2124976c909dd440bea4499c5fd9c8ca4da95fd8/data/packed/latest.json) as soon as [it has been initialized](https://github.com/moment/moment-timezone/blob/de5d1f8d17074af873366b2453b442552ddd35e4/index.js#L2).

## Upgrading

We are currently using the following versions of those modules:

* `moment`: 2.10.6
* `moment-timezone`: 0.4.0
* `i18n-calypso`: 1.7.1

This pull request (and [this one](https://github.com/Automattic/i18n-calypso/pull/31)) takes care of upgrading them to the latest available version:

* `moment`: 2.18.1
* `moment-timezone`: 0.5.11
* `i18n-calypso`: 1.7.4

## Analyzing

Running `make analyze-bundles` gives us the following metrics about the amount of data bundled with the new version of those modules:

Module            | Data parsed | Data gzipped
----------------- | ----------- | ------------
`moment`          | 327 KB      | 79KB
`moment-timezone` | 176 KB      | 23KB

<img width="956" alt="screenshot" src="https://cloud.githubusercontent.com/assets/594356/26067723/6a9e07c8-399b-11e7-9ed8-0b87026482d0.png">


## Understanding

We use `moment` directly (as opposed to [via `i18n-calypso`](https://github.com/Automattic/wp-calypso/blob/e002e8d6178abc77c7cbfcf3ad1a08a89e8c5208/client/me/account/main.jsx#L291)) in a lot of places, although not always in an [explicit way](https://github.com/Automattic/wp-calypso/blob/e002e8d6178abc77c7cbfcf3ad1a08a89e8c5208/client/me/account/main.jsx#L256). This is because of the [i18n mixin](https://github.com/Automattic/i18n-calypso/blob/63df0ed71cbd1dc57ea13152178b348b0f81e2f2/lib/mixin/index.js) that we [inject in all React components](https://github.com/Automattic/wp-calypso/blob/3e30a33926bd653d75396a9a348462ab99da4d34/client/boot/index.js#L72) and that exposes it automatically. The locale of `moment` is [set globally by `i18n-calypso`](https://github.com/Automattic/i18n-calypso/blob/e40d408e6e89f2738ac58415d98bc191254b5a19/lib/index.js#L193) to the user's locale during [client bootstrapping](https://github.com/Automattic/wp-calypso/blob/af79ae371e6d4d5814cc7aaef7eef5a8ec3e5285/client/boot/index.js).

## Solving

We can configure Webpack pretty easily to remove data for `moment` and `moment-timezone` from the `vendor` bundle. However loading this data on-demand is much more complicated.

In the case of `moment` - which always ships with the English locale - we must be able to load a new [locale file](https://github.com/moment/moment/tree/b8a7fc310eb3625e83fc0c8f1ea2840fa83c7378/locale) whenever the [locale changes](http://momentjs.com/docs/#/i18n/). In the case of `moment-timezone`, we must load the [data](https://github.com/moment/moment-timezone/blob/2124976c909dd440bea4499c5fd9c8ca4da95fd8/data/packed/latest.json) as soon as we want to use it (i.e. as soon as we call [`moment.tz` or `moment().tz`](http://momentjs.com/timezone/docs/#/using-timezones/)).

## Testing

1. Run `git checkout try/minify-moment` and start your server, or open a [live branch](https://calypso.live/?branch=try/minify-moment)
2. Log in to an English account
3. Check that the [`vendor` bundle](http://calypso.localhost:3000/calypso/vendor.HASH.js) is now ~1.7M (instead of ~1.9M) in the network panel of your browser
4. Check that no [`moment-locale`](http://calypso.localhost:3000/calypso/moment-locale-LOCALE.HASH.js) bundle was loaded in this panel
5. Navigate to the [`Account Settings` page](http://calypso.localhost:3000/me/account)
6. Notice the date in English just below the `Username` field (e.g. `Joined April 2016`) 
7. Select a different language in the `Interface Language` select box
8. Hit the `Save Account Settings` button, which will reload the page
9. Check that a `moment-locale`bundle was loaded this time in the network panel
10. Check that the date below the `Username` field is now in the language chosen
